### PR TITLE
RavenDB-21573 Do not change analyzer when field has time during QueryBuilder phase instead do it in CoraxBooleanItem.

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21573

### Additional description

We moved the mechanism of detecting dates into CoraxBooleanItem instead of the QueryBuilder phase.

### Type of change

- Bug fix

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs [hunk 1]
```diff
@@ -72,10 +88,16 @@ private MultiTermMatch InQuery<TTermType>(in FieldMetadata field, List<TTermType
         return MultiTermMatch.Create(new MultiTermMatch<InTermProvider<TTermType>>(this, field, _transaction.Allocator, new InTermProvider<TTermType>(this, field, inTerms), streamingEnabled: false, token: token));
     }
 
-    public IQueryMatch AllInQuery(string field, HashSet<string> allInTerms, in CancellationToken token = default) => AllInQuery(FieldMetadataBuilder(field), allInTerms, token: token);
+    // public IQueryMatch AllInQuery(string field, HashSet<string> allInTerms, in CancellationToken token = default)
+    // {
+    //     return default;
+    // } // => AllInQuery(FieldMetadataBuilder(field), allInTerms, token: token);
```
Comment (redknightlois) [inline src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs]: Remove this commented code.

File: src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs [hunk 2]
```diff
@@ -84,6 +106,7 @@ private MultiTermMatch InQuery<TTermType>(in FieldMetadata field, List<TTermType
     //In this case, when we get more conditions, we have to quit building the tree and manually check the entries with UnaryMatch.
     private IQueryMatch AllInQuery<TTerm>(in FieldMetadata field, HashSet<TTerm> allInTerms, bool skipEmptyItems = false, in CancellationToken cancellationToken = default)
     {
+        Debug.Assert(typeof(TTerm) == typeof(InParameter) || typeof(TTerm) == typeof(Slice) || typeof(TTerm) == typeof(string));
         const int maximumTermMatchesHandledAsTermMatches = 4;
```
Comment (redknightlois) [inline src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs:110]: Should throw with `UnsupportedException` if it is not in the list of valid cases (check will be evicted anyways by the JIT).

File: src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs [hunk 3]
```diff
@@ -105,7 +128,19 @@ private IQueryMatch AllInQuery<TTerm>(in FieldMetadata field, HashSet<TTerm> all
...
+            {
+                var casted = (InParameter)(object)item;
+                itemSlice = casted.Exact 
+                    ? EncodeAndApplyAnalyzer(default, casted.Term) 
+                    : EncodeAndApplyAnalyzer(field, casted.Term);
+            }
+            else if (typeof(TTerm) == typeof(string))
+                itemSlice = EncodeAndApplyAnalyzer(field, (string)(object)item);
+            else
+                itemSlice = (Slice)(object)item;
```
Comment (redknightlois) [inline src/Corax/Querying/MultiTermMatches/IndexSearcher.MultiTermMatches.In.cs:142]: Same comment, do the type check explicit and throw `UnsupportedException` here.

**Comment by redknightlois** [comment]
After comments addressed LGTM.

**Comment by maciejaszyk** [comment]
All done

**Pull request merged**